fabtests/prov/efa/src/efa_exhaust_mr_reg_rdm_pingpong.c

```diff
@@ -0,0 +1,124 @@
+/* Copyright Amazon.com, Inc. or its affiliates. All rights reserved. */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <getopt.h>
+
+#include <rdma/fi_errno.h>
+
+#include "shared.h"
+#include "benchmarks/benchmark_shared.h"
+#include "efa_exhaust_mr_reg_common.h"
+
+static int run(void)
+{
+	int i, ret = 0;
+
+	if (!(opts.options & FT_OPT_SIZE)) {
+		for (i = 0; i < TEST_CNT; i++) {
+			if (!ft_use_size(i, opts.sizes_enabled))
+				continue;
+			opts.transfer_size = test_size[i].size;
+			init_test(&opts, test_name, sizeof(test_name));
+			ret = pingpong();
+			if (ret)
+				return ret;
+		}
+	} else {
+		init_test(&opts, test_name, sizeof(test_name));
+		ret = pingpong();
+		if (ret)
+			return ret;
+	}
+
+	return ft_finalize();
+}
+
+int main(int argc, char **argv)
+{
+	int op, ret, err;
+	size_t registered;
+	void *buffers[EFA_MR_REG_LIMIT];
+	struct ibv_mr *mr_reg_vec[EFA_MR_REG_LIMIT];
+	struct ibv_pd *pd;
+
+	opts = INIT_OPTS;
+	opts.options |= FT_OPT_SKIP_REG_MR;
+	opts.mr_mode &= ~FI_MR_LOCAL;
+
+	hints = fi_allocinfo();
+	if (!hints)
+		return EXIT_FAILURE;
+
+	while ((op = getopt_long(argc, argv, "Uh" CS_OPTS INFO_OPTS BENCHMARK_OPTS,
+				 long_opts, &lopt_idx)) != -1) {
+		switch (op) {
+		default:
+			if (!ft_parse_long_opts(op, optarg))
+				continue;
+			ft_parse_benchmark_opts(op, optarg);
+			ft_parseinfo(op, optarg, hints, &opts);
+			ft_parsecsopts(op, optarg, &opts);
+			break;
+		case 'U':
+			hints->tx_attr->op_flags |= FI_DELIVERY_COMPLETE;
+			break;
+		case '?':
+		case 'h':
+			ft_csusage(argv[0], "Ping pong client and server using RDM after exhausting MR limits on the EFA device.");
+			ft_benchmark_usage();
+			ft_longopts_usage();
+			return EXIT_FAILURE;
+		}
+	}
+
+	if (optind < argc)
+		opts.dst_addr = argv[optind];
+
+	hints->ep_attr->type = FI_EP_RDM;
+	hints->caps = FI_MSG;
+	hints->mode |= FI_CONTEXT;
+	hints->domain_attr->mr_mode = opts.mr_mode;
+	hints->domain_attr->threading = FI_THREAD_DOMAIN;
+	hints->tx_attr->tclass = FI_TC_LOW_LATENCY;
+	hints->addr_format = opts.address_format;
+
+	ret = ft_init_fabric();
+	if (ret)
+		return ft_exit_code(ret);
+
+	ft_sync();
+	if (opts.dst_addr) {
+		err = ft_efa_alloc_bufs(buffers, EFA_MR_REG_BUF_SIZE,
+					EFA_MR_REG_LIMIT);
+		if (err)
+			FT_PRINTERR("alloc bufs", -err);
+
+		err = ft_efa_setup_ibv_pd(&pd);
+		if (err)
+			FT_PRINTERR("ibv protection domain", -err);
+
+		printf("Exhausting MRs on client\n");
+		err = ft_efa_register_mr_reg(pd, buffers, EFA_MR_REG_BUF_SIZE,
```

Comment:
If you drain the MR reg here, can the bounce buffer pool still able to be growed? I remember we currently grow the tx/rx pkt pool during the first fi_cq_read.

Reply:
Thanks for catching this. I thought the `ft_sync` before the MR exhaustion was exchanging message through EFA provider but it's actually through sockets because we ask for OOB address exchange. I added a progress call now.

During my tests, the ft_sync was calling progress because of this line but we're not always guaranteed to hit this line, so better to have an explicit progress call
https://github.com/ofiwg/libfabric/blob/faf4ff43ae5ab9c2e0f6f3b8a31b1d6bb142572e/fabtests/common/shared.c#L3794

Reply:
Yeah, you need to force progress before draining the MR, ft_sync can be oob if `-b` is used